lib/hiera/backend.rb

```diff
@@ -233,6 +249,16 @@ def lookup(key, default, scope, order_override, resolution_type)
       def clear!
         @backends = {}
       end
+
+      def qualified_lookup(segments, hash)
+        value = hash
+        segments.each do |segment|
+          break if value.nil? || value == :undefined
+          raise Exception, "Hiera type mismatch: can't index #{value.class.name} with '#{segment}'" unless value.kind_of? Hash
```

Comment:
Not sure if that use case was in scope. If it is, the simplest way to implement it is probably to say that integer keys are in fact zero based indexes into an array, i.e. `x.3` will require that `x` is an array and yield the fourth element. Assumes that no strings containing integers alone are used as hiera keys. If integer hash keys are used, then we need another notation such as [] which brings on the discussion about what can be used as an argument. Alternatives are:

* Only a literal integer (hash keys use dot notation)
* An literal integer or a String, i.e. the array index `x[3]` is different from the hash key `x['3']` (now we no longer need the dot notation)
* An expression, i.e.`x[%{key.to.lookup}]`

My preference would be to keep it simple and use dot notation only, i.e. either don't permit array index or use `x.3`.